File: shared/telemetry-types/python/events.py
"""
Wrack Telemetry — Shared Python event type definitions.

These types mirror the canonical JSON Schema definitions in
``shared/telemetry-types/schemas/``. Keep them in sync when schemas evolve.

Usage::

    from events import EventType, EventSource, VALID_EVENT_TYPES

"""
from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional, Union

# ---------------------------------------------------------------------------
# Enumerations (as string literals for micropython compatibility)
# ---------------------------------------------------------------------------

EventSource = Literal["ev3", "rpi", "cloud_functions", "web", "ios"]

VALID_SOURCES: List[str] = ["ev3", "rpi", "cloud_functions", "web", "ios"]

EventType = Literal[
    "battery_status",
    "command_received",
    "command_executed",
    "device_status",
    "error",
    "api_request",
    "motor_status",
    "sensor_reading",
    "terrain_scan",
    "connection_status",
]

VALID_EVENT_TYPES: List[str] = [
    "battery_status",
    "command_received",
    "command_executed",
    "device_status",
    "error",
    "api_request",
    "motor_status",
    "sensor_reading",
    "terrain_scan",
    "connection_status",
]

BatteryType = Literal["rechargeable", "alkaline", "unknown"]
ControllerType = Literal["ps4", "network_remote", "unknown"]
DeviceStatusValue = Literal["connected", "disconnected", "error", "stalled", "initializing"]
DeviceType = Literal["motor", "sensor", "controller", "unknown"]
HttpMethod = Literal["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"]

# ---------------------------------------------------------------------------
# Type aliases — using TypedDict where available, plain Dict otherwise
# ---------------------------------------------------------------------------

try:
    from typing import TypedDict

    class TelemetryEventEnvelope(TypedDict, total=False):
        """Common envelope shared by every telemetry event."""
        event_id: str               # required
        event_type: str             # required — one of VALID_EVENT_TYPES
        source: str                 # required — one of VALID_SOURCES
        timestamp: str              # required — ISO 8601 UTC
        payload: Dict[str, Any]     # required
        session_id: Optional[str]
        device_id: Optional[str]
        version: Optional[str]
        tags: Optional[List[str]]
        user_id: Optional[str]
        correlation_id: Optional[str]

    class BatteryStatusPayload(TypedDict, total=False):
        voltage_mv: int             # required
        percentage: float           # required, 0–100
        voltage_v: Optional[float]
        current_ma: Optional[float]
        battery_type: Optional[str]
        is_critical: Optional[bool]

    class CommandReceivedPayload(TypedDict, total=False):
        command: str                # required
        params: Optional[Dict[str, Any]]
        controller_type: Optional[str]
        received_at_ms: Optional[float]

    class CommandExecutedPayload(TypedDict, total=False):
        command: str                # required
        success: bool               # required
        duration_ms: Optional[float]
        error_message: Optional[str]
        params: Optional[Dict[str, Any]]
        controller_type: Optional[str]

    class DeviceStatusPayload(TypedDict, total=False):
        device_name: str            # required
        status: str                 # required — one of VALID_DEVICE_STATUSES
        device_type: Optional[str]
        port: Optional[str]
        previous_status: Optional[str]
        error_message: Optional[str]

    class ErrorPayload(TypedDict, total=False):
        error_type: str             # required
        message: str                # required
        error_code: Optional[str]
        component: Optional[str]
        stack_trace: Optional[str]
        context: Optional[Dict[str, Any]]

    class ApiRequestPayload(TypedDict, total=False):
        endpoint: str               # required
        status_code: int            # required
        latency_ms: float           # required
        method: Optional[str]
        command: Optional[str]
        robot_response_time_ms: Optional[float]
        client_ip_hash: Optional[str]
        error_message: Optional[str]

except ImportError:
    # MicroPython fallback — TypedDict is not available; use plain dicts
    TelemetryEventEnvelope = dict  # type: ignore[misc, assignment]
    BatteryStatusPayload = dict     # type: ignore[misc, assignment]
    CommandReceivedPayload = dict   # type: ignore[misc, assignment]
    CommandExecutedPayload = dict   # type: ignore[misc, assignment]
    DeviceStatusPayload = dict      # type: ignore[misc, assignment]
    ErrorPayload = dict             # type: ignore[misc, assignment]
    ApiRequestPayload = dict        # type: ignore[misc, assignment]


# ---------------------------------------------------------------------------
# Convenience constants
# ---------------------------------------------------------------------------

VALID_DEVICE_STATUSES: List[str] = [
    "connected", "disconnected", "error", "stalled", "initializing"
]

VALID_DEVICE_TYPES: List[str] = ["motor", "sensor", "controller", "unknown"]

VALID_BATTERY_TYPES: List[str] = ["rechargeable", "alkaline", "unknown"]

VALID_CONTROLLER_TYPES: List[str] = ["ps4", "network_remote", "unknown"]

VALID_HTTP_METHODS: List[str] = ["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"]

P0_EVENT_TYPES: List[str] = [
    "battery_status",
    "command_received",
    "command_executed",
    "device_status",
    "error",
    "api_request",
]

File: shared/telemetry-types/typescript/events.ts
/**
 * Wrack Telemetry — Shared TypeScript event types and validation helpers.
 *
 * These types mirror the canonical JSON Schema definitions in
 * `shared/telemetry-types/schemas/`. Keep them in sync when schemas evolve.
 *
 * Usage:
 *   import { TelemetryEvent, EventType, validateEventEnvelope } from './events';
 */

// ---------------------------------------------------------------------------
// Common envelope
// ---------------------------------------------------------------------------

export type EventSource = 'ev3' | 'rpi' | 'cloud_functions' | 'web' | 'ios';

export type EventType =
  | 'battery_status'
  | 'command_received'
  | 'command_executed'
  | 'device_status'
  | 'error'
  | 'api_request'
  | 'motor_status'
  | 'sensor_reading'
  | 'terrain_scan'
  | 'connection_status';

/** Common envelope shared by every telemetry event. */
export interface TelemetryEventEnvelope {
  event_id: string;
  event_type: EventType;
  source: EventSource;
  /** ISO 8601 UTC timestamp of when the event occurred at the source. */
  timestamp: string;
  session_id?: string | null;
  device_id?: string | null;
  payload: Record<string, unknown>;
  version?: string | null;
  tags?: string[] | null;
  user_id?: string | null;
  correlation_id?: string | null;
}

// ---------------------------------------------------------------------------
// P0 payload types
// ---------------------------------------------------------------------------

export type BatteryType = 'rechargeable' | 'alkaline' | 'unknown';

/** Payload for `battery_status` events. */
export interface BatteryStatusPayload {
  /** Battery voltage in millivolts. */
  voltage_mv: number;
  /** Battery voltage in volts. */
  voltage_v?: number;
  /** Battery current draw in milliamps. */
  current_ma?: number;
  /** Estimated remaining charge 0–100. */
  percentage: number;
  battery_type?: BatteryType;
  /** True when voltage is below the low-battery threshold. */
  is_critical?: boolean;
}

export type ControllerType = 'ps4' | 'network_remote' | 'unknown';

/** Payload for `command_received` events. */
export interface CommandReceivedPayload {
  command: string;
  params?: Record<string, unknown> | null;
  controller_type?: ControllerType;
  received_at_ms?: number;
}

/** Payload for `command_executed` events. */
export interface CommandExecutedPayload {
  command: string;
  success: boolean;
  duration_ms?: number;
  error_message?: string | null;
  params?: Record<string, unknown> | null;
  controller_type?: ControllerType;
}

export type DeviceStatusValue =
  | 'connected'
  | 'disconnected'
  | 'error'
  | 'stalled'
  | 'initializing';

export type DeviceType = 'motor' | 'sensor' | 'controller' | 'unknown';

/** Payload for `device_status` events. */
export interface DeviceStatusPayload {
  device_name: string;
  device_type?: DeviceType;
  port?: string | null;
  status: DeviceStatusValue;
  previous_status?: DeviceStatusValue | 'unknown' | null;
  error_message?: string | null;
}

/** Payload for `error` events. */
export interface ErrorPayload {
  error_type: string;
  error_code?: string | null;
  message: string;
  component?: string | null;
  stack_trace?: string | null;
  context?: Record<string, unknown> | null;
}

export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | 'OPTIONS';

/** Payload for `api_request` events. */
export interface ApiRequestPayload {
  endpoint: string;
  method?: HttpMethod;
  command?: string | null;
  status_code: number;
  latency_ms: number;
  robot_response_time_ms?: number | null;
  client_ip_hash?: string | null;
  error_message?: string | null;
}

// ---------------------------------------------------------------------------
// Discriminated union — typed event wrappers
// ---------------------------------------------------------------------------

export type BatteryStatusEvent = TelemetryEventEnvelope & {
  event_type: 'battery_status';
  payload: BatteryStatusPayload;
};

export type CommandReceivedEvent = TelemetryEventEnvelope & {
  event_type: 'command_received';
  payload: CommandReceivedPayload;
};

export type CommandExecutedEvent = TelemetryEventEnvelope & {
  event_type: 'command_executed';
  payload: CommandExecutedPayload;
};

export type DeviceStatusEvent = TelemetryEventEnvelope & {
  event_type: 'device_status';
  payload: DeviceStatusPayload;
};

export type ErrorEvent = TelemetryEventEnvelope & {
  event_type: 'error';
  payload: ErrorPayload;
};

export type ApiRequestEvent = TelemetryEventEnvelope & {
  event_type: 'api_request';
  payload: ApiRequestPayload;
};

/** Union of all typed telemetry events. */
export type TelemetryEvent =
  | BatteryStatusEvent
  | CommandReceivedEvent
  | CommandExecutedEvent
  | DeviceStatusEvent
  | ErrorEvent
  | ApiRequestEvent;

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

/** Errors collected during validation. */
export interface ValidationError {
  field: string;
  message: string;
}

export interface ValidationResult {
  valid: boolean;
  errors: ValidationError[];
}

const VALID_SOURCES: readonly EventSource[] = [
  'ev3', 'rpi', 'cloud_functions', 'web', 'ios',
] as const;

const VALID_EVENT_TYPES: readonly EventType[] = [
  'battery_status', 'command_received', 'command_executed',
  'device_status', 'error', 'api_request',
  'motor_status', 'sensor_reading', 'terrain_scan', 'connection_status',
] as const;

const ISO_8601_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z$/;
const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

/**
 * Validates the common event envelope fields.
 * Does NOT validate the `payload` — call the type-specific validator for that.
 */
export function validateEventEnvelope(
  event: unknown,
): ValidationResult {
  const errors: ValidationError[] = [];

  if (typeof event !== 'object' || event === null) {
    return { valid: false, errors: [{ field: 'root', message: 'Event must be a non-null object' }] };
  }

  const e = event as Record<string, unknown>;

  if (typeof e.event_id !== 'string' || !UUID_RE.test(e.event_id)) {
    errors.push({ field: 'event_id', message: 'Must be a valid UUID v4 string' });
  }

  if (typeof e.event_type !== 'string' || !(VALID_EVENT_TYPES as readonly string[]).includes(e.event_type)) {
    errors.push({
      field: 'event_type',
      message: `Must be one of: ${VALID_EVENT_TYPES.join(', ')}`,
    });
  }

  if (typeof e.source !== 'string' || !(VALID_SOURCES as readonly string[]).includes(e.source)) {
    errors.push({
      field: 'source',
      message: `Must be one of: ${VALID_SOURCES.join(', ')}`,
    });
  }

  if (typeof e.timestamp !== 'string' || !ISO_8601_RE.test(e.timestamp)) {
    errors.push({
      field: 'timestamp',
      message: 'Must be an ISO 8601 UTC timestamp string ending in Z (e.g. 2026-01-01T00:00:00Z)',
    });
  }

  if (typeof e.payload !== 'object' || e.payload === null || Array.isArray(e.payload)) {
    errors.push({ field: 'payload', message: 'Must be a non-null object' });
  }

  return { valid: errors.length === 0, errors };
}

/** Validates a `battery_status` payload. */
export function validateBatteryStatusPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.voltage_mv !== 'number' || !Number.isInteger(p.voltage_mv) || p.voltage_mv < 0) {
    errors.push({ field: 'payload.voltage_mv', message: 'Must be a non-negative integer' });
  }
  if (typeof p.percentage !== 'number' || p.percentage < 0 || p.percentage > 100) {
    errors.push({ field: 'payload.percentage', message: 'Must be a number between 0 and 100' });
  }
  if (p.voltage_v !== undefined && p.voltage_v !== null && (typeof p.voltage_v !== 'number' || p.voltage_v < 0)) {
    errors.push({ field: 'payload.voltage_v', message: 'Must be a non-negative number' });
  }
  if (p.is_critical !== undefined && p.is_critical !== null && typeof p.is_critical !== 'boolean') {
    errors.push({ field: 'payload.is_critical', message: 'Must be a boolean' });
  }

  return { valid: errors.length === 0, errors };
}

/** Validates a `command_received` payload. */
export function validateCommandReceivedPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.command !== 'string' || p.command.trim() === '') {
    errors.push({ field: 'payload.command', message: 'Must be a non-empty string' });
  }

  return { valid: errors.length === 0, errors };
}

/** Validates a `command_executed` payload. */
export function validateCommandExecutedPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.command !== 'string' || p.command.trim() === '') {
    errors.push({ field: 'payload.command', message: 'Must be a non-empty string' });
  }
  if (typeof p.success !== 'boolean') {
    errors.push({ field: 'payload.success', message: 'Must be a boolean' });
  }
  if (p.duration_ms !== undefined && p.duration_ms !== null &&
      (typeof p.duration_ms !== 'number' || p.duration_ms < 0)) {
    errors.push({ field: 'payload.duration_ms', message: 'Must be a non-negative number' });
  }

  return { valid: errors.length === 0, errors };
}

const VALID_DEVICE_STATUSES = ['connected', 'disconnected', 'error', 'stalled', 'initializing'] as const;

/** Validates a `device_status` payload. */
export function validateDeviceStatusPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.device_name !== 'string' || p.device_name.trim() === '') {
    errors.push({ field: 'payload.device_name', message: 'Must be a non-empty string' });
  }
  if (typeof p.status !== 'string' || !(VALID_DEVICE_STATUSES as readonly string[]).includes(p.status)) {
    errors.push({
      field: 'payload.status',
      message: `Must be one of: ${VALID_DEVICE_STATUSES.join(', ')}`,
    });
  }

  return { valid: errors.length === 0, errors };
}

/** Validates an `error` payload. */
export function validateErrorPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.error_type !== 'string' || p.error_type.trim() === '') {
    errors.push({ field: 'payload.error_type', message: 'Must be a non-empty string' });
  }
  if (typeof p.message !== 'string' || p.message.trim() === '') {
    errors.push({ field: 'payload.message', message: 'Must be a non-empty string' });
  }

  return { valid: errors.length === 0, errors };
}

/** Validates an `api_request` payload. */
export function validateApiRequestPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.endpoint !== 'string' || p.endpoint.trim() === '') {
    errors.push({ field: 'payload.endpoint', message: 'Must be a non-empty string' });
  }
  if (typeof p.status_code !== 'number' || !Number.isInteger(p.status_code) ||
      p.status_code < 100 || p.status_code > 599) {
    errors.push({ field: 'payload.status_code', message: 'Must be an integer HTTP status code (100–599)' });
  }
  if (typeof p.latency_ms !== 'number' || p.latency_ms < 0) {
    errors.push({ field: 'payload.latency_ms', message: 'Must be a non-negative number' });
  }

  return { valid: errors.length === 0, errors };
}

// ---------------------------------------------------------------------------
// Convenience: validate envelope + payload together
// ---------------------------------------------------------------------------

const PAYLOAD_VALIDATORS: Partial<Record<EventType, (p: unknown) => ValidationResult>> = {
  battery_status: validateBatteryStatusPayload,
  command_received: validateCommandReceivedPayload,
  command_executed: validateCommandExecutedPayload,
  device_status: validateDeviceStatusPayload,
  error: validateErrorPayload,
  api_request: validateApiRequestPayload,
};

/**
 * Validates the full event (envelope + payload).
 * Errors from both layers are merged into a single result.
 */
export function validateTelemetryEvent(event: unknown): ValidationResult {
  const envelopeResult = validateEventEnvelope(event);

  if (!envelopeResult.valid) {
    return envelopeResult;
  }

  const e = event as TelemetryEventEnvelope;
  const payloadValidator = PAYLOAD_VALIDATORS[e.event_type];

  if (!payloadValidator) {
    return { valid: true, errors: [] };
  }

  const payloadResult = payloadValidator(e.payload);
  return {
    valid: payloadResult.valid,
    errors: payloadResult.errors,
  };
}
